Add GridMap octant related queries and tests

smix8 opened the pull request:
Adds GridMap octant queries.

Requires https://github.com/godotengine/godot/pull/105329 to be merged first.

Adds octant query functions as a performance alternative to the existing cell functions.

E.g. when only a few cells for some specific octants are needed, e.g. inside some AABB bounds, and not all the cells of the entire GridMap which creates potentially a slow array with thousands of entries.

`TypedArray[Vector3i] get_used_octants()`
Returns the octant coordinates as Vector3i of all octants that have at least 1 cell in use.

`TypedArray[Vector3i] get_used_octants_by_item( p_item: int)`
Returns the used octant coordinates of all octants that have at least 1 cell with the specific cell item in use.

`TypedArray[Vector3i] get_used_cells_in_octant(p_octant_coords: Vector3i)`
Returns the used cell coordinates of all the cells inside the specific octant coordinate.

`TypedArray[Vector3i] get_used_cells_in_octant_by_item(p_octant_coords: Vector3i, p_item: int)`
Returns the used cell coordinates of all the cells inside the specific octant coordinate that have the specific cell item in use.

`TypedArray[Vector3i] get_used_octants_in_bounds(p_bounds: AABB)`
Returns the used octant coordinates of all octants that have at least 1 cell used with the specific local bounds.

`Vector3i get_octant_coords_from_cell_coords(p_cell_coords: Vector3i)`
Script helper to convert cell coordinate to octant coordinate for the octant related query functions.

<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

smix8 committed:
Add GridMap octant queries and tests

Adds GridMap octant queries and tests.

rburing reviewed (changes requested):
Functionality looks good, and I would say the debug PR already justifies adding this API. I have just some comments on the docs.

smix8 commented:
I realized that I needed 2 native functions that are not slowed down by `Variant` / `TypedArray`. Need them to speed up e.g. navmesh parsing or editor plugin. So I added I added `get_index_keys_in_bounds()` and `get_octant_keys_in_bounds()`.

Realizing further that so many new queries will turn into a nice bug surface I also added some basic GridMap tests.

kiroxas commented on modules/gridmap/grid_map.cpp:
```diff
@@ -547,6 +547,123 @@ GridMap::OctantKey GridMap::get_octant_key_from_cell_coords(const Vector3i &p_ce
 }
 
+Vector3i GridMap::get_octant_coords_from_cell_coords(const Vector3i &p_cell_coords) const {
+	return Vector3i(
```
Maybe cache `float sizeMinusOne = octant_size - 1;` and `float inverseSize = 1.f / octant_size;`  may be shorter and more readable, and debug builds will not do 3 divs

smix8 replied:
One point in the review further up was to remove the use of float. This now is a straight copy of the TileMap cell raster code. If it makes sense to change it again to a cached version for performance should preferably be its own PR with performance tests that change both.

Mickeon commented on modules/gridmap/doc_classes/GridMap.xml:
```diff
@@ -96,6 +96,20 @@
+			<return type="Vector3i[]" />
+			<param index="0" name="bounds" type="AABB" />
+			<description>
+				Returns an array of [Vector3i] octant coordinates that fall inside the [param bounds], including octants that have no cells in use.
```
```suggestion
				Returns an array of [Vector3i] octant coordinates that are inside the given [param bounds], including octants that have no cells in use.
```
Is it necessary to specify "**[Vector3i]** octant coordinates" every time?

smix8 replied:
Yeah, else the universe could fall apart. More serious, what is the concern / harm?

Mickeon replied:
It just doesn't feel right to use it as a shorter form of "__Vector3i representing the octant coordinates_". Plus the return type can be inferred from the function's signature and it's not particularly unusual.

Mickeon commented on modules/gridmap/doc_classes/GridMap.xml:
```diff
@@ -96,6 +96,20 @@
+			<return type="Vector3i" />
+			<param index="0" name="cell_coords" type="Vector3i" />
+			<description>
+				Returns the [Vector3i] octant coordinates of the octant that the cell with cell coordinates [param cell_coords] belongs to.
```
Feel free to disregard if you think this is worse:
```suggestion
				Returns the [Vector3i] octant coordinates of the octant that the cell at [param cell_coords] belongs to.
```

AThousandShips commented on modules/gridmap/grid_map.h:
```diff
@@ -301,6 +301,21 @@ class GridMap : public Node3D {
+
+	// Fastpath functions for native modules that do not use Variant / TypedArray.
+	LocalVector<IndexKey> get_index_keys_in_bounds(const AABB &p_bounds, bool p_used_only) const;
+	LocalVector<OctantKey> get_octant_keys_in_bounds(const AABB &p_bounds, bool p_used_only) const;
```
```suggestion
	LocalVector<IndexKey> get_index_keys_in_bounds(const AABB &p_bounds, bool p_used_only = true) const;
	LocalVector<OctantKey> get_octant_keys_in_bounds(const AABB &p_bounds, bool p_used_only = true) const;
```

AThousandShips commented on modules/gridmap/doc_classes/GridMap.xml:
```diff
@@ -116,6 +130,41 @@
+			<return type="Vector3i[]" />
+			<param index="0" name="octant_coords" type="Vector3i" />
+			<description>
+				Returns an array of [Vector3i] with the cell coordinates of non-empty cells inside the octant at [param octant_coords].
```
```suggestion
				Returns an array of [Vector3i]s with the cell coordinates of non-empty cells inside the octant at [param octant_coords].
```

AThousandShips commented on modules/gridmap/doc_classes/GridMap.xml:
```diff
@@ -116,6 +130,41 @@
+			<param index="0" name="octant_coords" type="Vector3i" />
+			<param index="1" name="item" type="int" />
+			<description>
+				Returns an array of [Vector3i] with the cell coordinates of cells inside the octant at [param octant_coords] that use the specified cell [param item].
```
```suggestion
				Returns an array of [Vector3i]s with the cell coordinates of cells inside the octant at [param octant_coords] that use the specified cell [param item].
```

AThousandShips commented on modules/gridmap/doc_classes/GridMap.xml:
```diff
@@ -116,6 +130,41 @@
+		<method name="get_used_octants" qualifiers="const">
+			<return type="Vector3i[]" />
+			<description>
+				Returns an array of [Vector3i] with the octant coordinates of the non-empty octants in the grid map.
```
```suggestion
				Returns an array of [Vector3i]s with the octant coordinates of the non-empty octants in the grid map.
```

AThousandShips commented on modules/gridmap/grid_map.cpp:
```diff
@@ -550,6 +550,123 @@ GridMap::OctantKey GridMap::get_octant_key_from_cell_coords(const Vector3i &p_ce
+			p_cell_coords.z > 0 ? p_cell_coords.z / octant_size : (p_cell_coords.z - (octant_size - 1)) / octant_size);
+}
+
+LocalVector<GridMap::IndexKey> GridMap::get_index_keys_in_bounds(const AABB &p_bounds, bool p_used_only = true) const {
```
```suggestion
LocalVector<GridMap::IndexKey> GridMap::get_index_keys_in_bounds(const AABB &p_bounds, bool p_used_only) const {
```
